File: packages/server/src/routes/tasks.ts
import type { Hono } from 'hono'
import { OpenAPIHono, createRoute, z } from '@hono/zod-openapi'
import type { Context } from 'hono'
import { checkScope } from '../auth.js'
import { getSubscriberCount } from './sse.js'
import type { SubscriberCounts } from './sse.js'
import {
  CreateTaskSchema,
  PublishEventSchema,
  TransitionSchema,
  TaskSchema,
  TaskEventSchema,
  ErrorSchema,
} from '../schemas.js'
import { TaskConflictError, InvalidTransitionError, SeqStaleError, SeqDuplicateError, SeqGapError, SeqTimeoutError, collapseAccumulateSeries } from '@taskcast/core'
import type { TaskEngine, CreateTaskInput, PublishEventInput, SinceCursor, TaskError, BlockedRequest, TaskFilter, TaskStatus, EventQueryOptions } from '@taskcast/core'

// ─── Route Definitions ─────────────────────────────────────────────────────

const createTaskRoute = createRoute({
  method: 'post',
  path: '/',
  tags: ['Tasks'],
  summary: 'Create a new task',
  security: [{ Bearer: [] }],
  request: {
    body: { content: { 'application/json': { schema: CreateTaskSchema } } },
  },
  responses: {
    201: { description: 'Task created', content: { 'application/json': { schema: TaskSchema } } },
    400: { description: 'Validation error', content: { 'application/json': { schema: ErrorSchema } } },
    403: { description: 'Forbidden', content: { 'application/json': { schema: ErrorSchema } } },
    409: { description: 'Task ID already exists', content: { 'application/json': { schema: ErrorSchema } } },
  },
})

const listTasksRoute = createRoute({
  method: 'get',
  path: '/',
  tags: ['Tasks'],
  summary: 'List tasks',
  description: 'List tasks with optional status and type filters.',
  security: [{ Bearer: [] }],
  request: {
    query: z.object({
      status: z.string().optional().openapi({ description: 'Comma-separated status filter' }),
      type: z.string().optional().openapi({ description: 'Task type filter' }),
    }),
  },
  responses: {
    200: { description: 'Task list', content: { 'application/json': { schema: z.object({ tasks: z.array(TaskSchema) }) } } },
    403: { description: 'Forbidden', content: { 'application/json': { schema: ErrorSchema } } },
  },
})

const getTaskRoute = createRoute({
  method: 'get',
  path: '/{taskId}',
  tags: ['Tasks'],
  summary: 'Get task by ID',
  security: [{ Bearer: [] }],
  request: {
    params: z.object({ taskId: z.string() }),
  },
  responses: {
    200: { description: 'Task details', content: { 'application/json': { schema: TaskSchema } } },
    404: { description: 'Task not found', content: { 'application/json': { schema: ErrorSchema } } },
    403: { description: 'Forbidden', content: { 'application/json': { schema: ErrorSchema } } },
  },
})

const transitionRoute = createRoute({
  method: 'patch',
  path: '/{taskId}/status',
  tags: ['Tasks'],
  summary: 'Transition task status',
  security: [{ Bearer: [] }],
  request: {
    params: z.object({ taskId: z.string() }),
    body: { content: { 'application/json': { schema: TransitionSchema } } },
  },
  responses: {
    200: { description: 'Updated task', content: { 'application/json': { schema: TaskSchema } } },
    400: { description: 'Validation error', content: { 'application/json': { schema: ErrorSchema } } },
    404: { description: 'Task not found', content: { 'application/json': { schema: ErrorSchema } } },
    403: { description: 'Forbidden', content: { 'application/json': { schema: ErrorSchema } } },
    409: { description: 'Invalid state transition', content: { 'application/json': { schema: ErrorSchema } } },
  },
})

const publishEventsRoute = createRoute({
  method: 'post',
  path: '/{taskId}/events',
  tags: ['Events'],
  summary: 'Publish events to a task',
  description: 'Supports single event or batch (array) publishing.',
  security: [{ Bearer: [] }],
  request: {
    params: z.object({ taskId: z.string() }),
    body: { content: { 'application/json': { schema: z.union([PublishEventSchema, z.array(PublishEventSchema)]) } } },
  },
  responses: {
    201: { description: 'Events published', content: { 'application/json': { schema: z.union([TaskEventSchema, z.array(TaskEventSchema)]) } } },
    400: { description: 'Validation error', content: { 'application/json': { schema: ErrorSchema } } },
    404: { description: 'Task not found', content: { 'application/json': { schema: ErrorSchema } } },
    403: { description: 'Forbidden', content: { 'application/json': { schema: ErrorSchema } } },
  },
})

const eventHistoryRoute = createRoute({
  method: 'get',
  path: '/{taskId}/events/history',
  tags: ['Events'],
  summary: 'Query event history',
  security: [{ Bearer: [] }],
  request: {
    params: z.object({ taskId: z.string() }),
    query: z.object({
      'since.id': z.string().optional(),
      'since.index': z.string().optional(),
      'since.timestamp': z.string().optional(),
      types: z.string().optional(),
      levels: z.string().optional(),
      limit: z.string().optional().openapi({ description: 'Maximum number of events to return' }),
      seriesFormat: z.string().optional().openapi({ description: 'Series format: delta (default) or accumulated' }),
    }),
  },
  responses: {
    200: { description: 'Event list', content: { 'application/json': { schema: z.array(TaskEventSchema) } } },
    404: { description: 'Task not found', content: { 'application/json': { schema: ErrorSchema } } },
    403: { description: 'Forbidden', content: { 'application/json': { schema: ErrorSchema } } },
  },
})

// ─── Router Factory ────────────────────────────────────────────────────────

// Engine types (Task, TaskEvent) are structurally compatible with the Zod schema
// output types but not identical. @hono/zod-openapi enforces strict return-type
// checking that cannot be satisfied without duplicating core types as Zod schemas.
// We bind openapi() with a relaxed signature to retain runtime route-spec
// registration and request validation while avoiding redundant type gymnastics.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type OpenAPIRegister = (route: any, handler: (c: Context) => Promise<Response>) => void

export function createTasksRouter(engine: TaskEngine, subscriberCounts: SubscriberCounts): Hono {
  const router = new OpenAPIHono()
  const register = router.openapi.bind(router) as OpenAPIRegister

  register(createTaskRoute, async (c) => {
    const auth = c.get('auth')
    if (!checkScope(auth, 'task:create')) return c.json({ error: 'Forbidden' }, 403)

    const body = await c.req.json()
    const parsed = CreateTaskSchema.safeParse(body)
    if (!parsed.success) return c.json({ error: parsed.error.flatten() }, 400)

    const d = parsed.data
    const input: CreateTaskInput = {}
    if (d.id !== undefined) input.id = d.id
    if (d.type !== undefined) input.type = d.type
    if (d.params !== undefined) input.params = d.params
    if (d.metadata !== undefined) input.metadata = d.metadata
    if (d.ttl !== undefined) input.ttl = d.ttl
    if (d.tags !== undefined) input.tags = d.tags
    if (d.assignMode !== undefined) input.assignMode = d.assignMode
    if (d.cost !== undefined) input.cost = d.cost
    if (d.disconnectPolicy !== undefined) input.disconnectPolicy = d.disconnectPolicy
    if (d.webhooks !== undefined) input.webhooks = d.webhooks as CreateTaskInput['webhooks']
    if (d.cleanup !== undefined) input.cleanup = d.cleanup as CreateTaskInput['cleanup']
    if (d.authConfig !== undefined) input.authConfig = d.authConfig as unknown as CreateTaskInput['authConfig']

    try {
      const task = await engine.createTask(input)
      return c.json(task, 201)
    } catch (err) {
      if (err instanceof TaskConflictError) return c.json({ error: err.message }, 409)
      const msg = err instanceof Error ? err.message : String(err)
      if (msg.includes('Invalid TTL') || msg.includes('Invalid cost')) return c.json({ error: msg }, 400)
      throw err
    }
  })

  register(listTasksRoute, async (c) => {
    const auth = c.get('auth')
    if (!checkScope(auth, 'event:subscribe')) return c.json({ error: 'Forbidden' }, 403)

    const status = c.req.query('status')
    const type = c.req.query('type')

    const filter: TaskFilter = {}
    if (status) filter.status = status.split(',').filter(Boolean) as TaskStatus[]
    if (type) filter.types = [type]

    const tasks = await engine.listTasks(filter)
    const enriched = tasks.map(t => {
      const subscriberCount = getSubscriberCount(subscriberCounts, t.id)
      return { ...t, hot: subscriberCount > 0, subscriberCount }
    })

    return c.json({ tasks: enriched })
  })

  register(getTaskRoute, async (c) => {
    const taskId = c.req.param('taskId') as string
    const auth = c.get('auth')
    if (!checkScope(auth, 'event:subscribe', taskId)) return c.json({ error: 'Forbidden' }, 403)

    const task = await engine.getTask(taskId)
    if (!task) return c.json({ error: 'Task not found' }, 404)
    const subscriberCount = getSubscriberCount(subscriberCounts, taskId)
    return c.json({ ...task, hot: subscriberCount > 0, subscriberCount })
  })

  register(transitionRoute, async (c) => {
    const taskId = c.req.param('taskId') as string
    const auth = c.get('auth')
    if (!checkScope(auth, 'task:manage', taskId)) return c.json({ error: 'Forbidden' }, 403)

    const body = await c.req.json()
    const parsed = TransitionSchema.safeParse(body)
    if (!parsed.success) return c.json({ error: parsed.error.flatten() }, 400)

    try {
      const payload: {
        result?: Record<string, unknown>
        error?: TaskError
        reason?: string
        ttl?: number
        resumeAfterMs?: number
        blockedRequest?: BlockedRequest
      } = {}
      if (parsed.data.result !== undefined) payload.result = parsed.data.result
      if (parsed.data.error !== undefined) {
        const e = parsed.data.error
        const taskError: TaskError = { message: e.message }
        if (e.code !== undefined) taskError.code = e.code
        if (e.details !== undefined) taskError.details = e.details
        payload.error = taskError
      }
      if (parsed.data.reason !== undefined) payload.reason = parsed.data.reason
      if (parsed.data.ttl !== undefined) payload.ttl = parsed.data.ttl
      if (parsed.data.resumeAfterMs !== undefined) payload.resumeAfterMs = parsed.data.resumeAfterMs
      if (parsed.data.blockedRequest !== undefined) {
        payload.blockedRequest = {
          type: parsed.data.blockedRequest.type,
          data: parsed.data.blockedRequest.data,
        }
      }
      const task = await engine.transitionTask(taskId, parsed.data.status, payload)
      return c.json(task)
    } catch (err) {
      if (err instanceof InvalidTransitionError) return c.json({ error: err.message }, 409)
      const msg = err instanceof Error ? err.message : String(err)
      if (msg.toLowerCase().includes('not found')) return c.json({ error: msg }, 404)
      return c.json({ error: msg }, 400)
    }
  })

  register(publishEventsRoute, async (c) => {
    const taskId = c.req.param('taskId') as string
    const auth = c.get('auth')
    if (!checkScope(auth, 'event:publish', taskId)) return c.json({ error: 'Forbidden' }, 403)

    const body = await c.req.json()
    const isBatch = Array.isArray(body)
    const inputs = isBatch ? body : [body]

    const events = []
    for (const input of inputs) {
      const parsed = PublishEventSchema.safeParse(input)
      if (!parsed.success) return c.json({ error: parsed.error.flatten() }, 400)
      try {
        const d = parsed.data
        const eventInput: PublishEventInput = { type: d.type, level: d.level, data: d.data }
        if (d.seriesId !== undefined) eventInput.seriesId = d.seriesId
        if (d.seriesMode !== undefined) eventInput.seriesMode = d.seriesMode
        if (d.seriesAccField !== undefined) eventInput.seriesAccField = d.seriesAccField
        if (d.clientId !== undefined) eventInput.clientId = d.clientId
        if (d.clientSeq !== undefined) eventInput.clientSeq = d.clientSeq
        if (d.seqMode !== undefined) eventInput.seqMode = d.seqMode
        const event = await engine.publishEvent(taskId, eventInput)
        events.push(event)
      } catch (err) {
        if (err instanceof SeqStaleError) {
          return c.json({ error: 'seq_stale', expectedSeq: err.expectedSeq, receivedSeq: err.receivedSeq }, 409)
        }
        if (err instanceof SeqDuplicateError) {
          return c.json({ error: 'seq_duplicate', seq: err.seq }, 409)
        }
        if (err instanceof SeqGapError) {
          return c.json({ error: 'seq_gap', expectedSeq: err.expectedSeq, receivedSeq: err.receivedSeq }, 409)
        }
        if (err instanceof SeqTimeoutError) {
          return c.json({ error: 'seq_timeout', seq: err.seq, expectedSeq: err.expectedSeq }, 408)
        }
        const msg = err instanceof Error ? err.message : String(err)
        if (msg.toLowerCase().includes('not found')) return c.json({ error: msg }, 404)
        return c.json({ error: msg }, 400)
      }
    }

    return c.json(isBatch ? events : events[0], 201)
  })

  register(eventHistoryRoute, async (c) => {
    const taskId = c.req.param('taskId') as string
    const auth = c.get('auth')
    if (!checkScope(auth, 'event:history', taskId)) return c.json({ error: 'Forbidden' }, 403)

    const task = await engine.getTask(taskId)
    if (!task) return c.json({ error: 'Task not found' }, 404)

    const sinceIndex = c.req.query('since.index')
    const sinceTimestamp = c.req.query('since.timestamp')
    const sinceId = c.req.query('since.id')
    const limitStr = c.req.query('limit')
    const seriesFormat = c.req.query('seriesFormat') ?? 'delta'

    let since: SinceCursor | undefined
    if (sinceId !== undefined || sinceIndex !== undefined || sinceTimestamp !== undefined) {
      since = {}
      if (sinceId !== undefined) since.id = sinceId
      if (sinceIndex !== undefined) since.index = Number(sinceIndex)
      if (sinceTimestamp !== undefined) since.timestamp = Number(sinceTimestamp)
    }

    const limitRaw = limitStr !== undefined ? parseInt(limitStr, 10) : undefined
    const limit = limitRaw !== undefined && Number.isFinite(limitRaw) && limitRaw > 0 ? limitRaw : undefined
    const opts: EventQueryOptions | undefined =
      since !== undefined || limit !== undefined
        ? { ...(since !== undefined && { since }), ...(limit !== undefined && { limit }) }
        : undefined

    let events = await engine.getEvents(taskId, opts)

    if (seriesFormat === 'accumulated') {
      events = await collapseAccumulateSeries(
        events,
        (tid, sid) => engine.getSeriesLatest(tid, sid),
      )
    }

    return c.json(events)
  })

  // ─── GET /tasks/:taskId/seq/:clientId — Query expected seq ─────────────────
  router.get('/:taskId/seq/:clientId', async (c: Context) => {
    const taskId = c.req.param('taskId') as string
    const clientId = c.req.param('clientId') as string
    const auth = c.get('auth')
    if (!checkScope(auth, 'event:publish', taskId)) return c.json({ error: 'Forbidden' }, 403)

    const expectedSeq = await engine.getExpectedSeq(taskId, clientId)
    if (expectedSeq === null) return c.json({ error: 'Client not initialized' }, 404)

    return c.json({ clientId, expectedSeq })
  })

  // ─── POST /tasks/:taskId/resolve — Resolve a blocked task ─────────────────
  router.post('/:taskId/resolve', async (c: Context) => {
    const taskId = c.req.param('taskId') as string
    const auth = c.get('auth')
    if (!checkScope(auth, 'task:resolve', taskId)) return c.json({ error: 'Forbidden' }, 403)

    const task = await engine.getTask(taskId)
    if (!task) return c.json({ error: 'Task not found' }, 404)
    if (task.status !== 'blocked') return c.json({ error: 'Task is not blocked' }, 400)

    const body = await c.req.json()
    const schema = z.object({ data: z.unknown() })
    const parsed = schema.safeParse(body)
    if (!parsed.success) return c.json({ error: parsed.error.flatten() }, 400)

    try {
      const updated = await engine.transitionTask(taskId, 'running', {
        result: typeof parsed.data.data === 'object' && parsed.data.data !== null
          ? parsed.data.data as Record<string, unknown>
          : { resolution: parsed.data.data },
      })
      return c.json(updated)
    } catch (err) {
      if (err instanceof InvalidTransitionError) return c.json({ error: err.message }, 409)
      const msg = err instanceof Error ? err.message : String(err)
      return c.json({ error: msg }, 400)
    }
  })

  // ─── GET /tasks/:taskId/request — Get blocked request ────────────────────
  router.get('/:taskId/request', async (c: Context) => {
    const taskId = c.req.param('taskId') as string
    const auth = c.get('auth')
    if (!checkScope(auth, 'task:resolve', taskId)) return c.json({ error: 'Forbidden' }, 403)

    const task = await engine.getTask(taskId)
    if (!task) return c.json({ error: 'Task not found' }, 404)
    if (task.status !== 'blocked' || !task.blockedRequest) {
      return c.json({ error: 'No blocked request' }, 404)
    }

    return c.json(task.blockedRequest)
  })

  return router as unknown as Hono
}

File: rust/taskcast-server/src/routes/tasks.rs
use std::collections::HashMap;
use std::sync::Arc;

use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::Extension;
use serde::Deserialize;
use serde_json::json;
use taskcast_core::{
    AssignMode, BlockedRequest, CleanupConfig, CreateTaskInput, DisconnectPolicy, EngineError,
    EventQueryOptions, Level, PublishEventInput, SeriesMode, SinceCursor, TaskAuthConfig,
    TaskEngine, TaskError, TaskFilter, TaskStatus, TransitionPayload, WebhookConfig,
};

use crate::auth::{check_scope, AuthContext};
use crate::error::AppError;
use crate::routes::sse::{get_subscriber_count, SubscriberCounts};

// ─── Request Bodies ──────────────────────────────────────────────────────────

#[derive(Debug, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct CreateTaskBody {
    pub id: Option<String>,
    pub r#type: Option<String>,
    pub params: Option<HashMap<String, serde_json::Value>>,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
    pub ttl: Option<u64>,
    pub webhooks: Option<Vec<WebhookConfig>>,
    pub cleanup: Option<CleanupConfig>,
    pub auth_config: Option<TaskAuthConfig>,
    pub tags: Option<Vec<String>>,
    pub assign_mode: Option<AssignMode>,
    pub cost: Option<u32>,
    pub disconnect_policy: Option<DisconnectPolicy>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct TransitionBody {
    pub status: TaskStatus,
    pub result: Option<HashMap<String, serde_json::Value>>,
    pub error: Option<TaskErrorBody>,
    pub reason: Option<String>,
    pub ttl: Option<u64>,
    pub resume_after_ms: Option<f64>,
    pub blocked_request: Option<BlockedRequest>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct TaskErrorBody {
    pub code: Option<String>,
    pub message: String,
    pub details: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PublishEventBody {
    pub r#type: String,
    pub level: Level,
    pub data: serde_json::Value,
    pub series_id: Option<String>,
    pub series_mode: Option<SeriesMode>,
    pub series_acc_field: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
pub enum EventsBody {
    Batch(Vec<PublishEventBody>),
    Single(PublishEventBody),
}

#[derive(Debug, Deserialize, utoipa::IntoParams)]
pub struct HistoryQuery {
    #[serde(rename = "since.index")]
    pub since_index: Option<u64>,
    #[serde(rename = "since.timestamp")]
    pub since_timestamp: Option<f64>,
    #[serde(rename = "since.id")]
    pub since_id: Option<String>,
    pub limit: Option<u64>,
    #[serde(rename = "seriesFormat")]
    pub series_format: Option<String>,
}

// ─── List Query ──────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize, utoipa::IntoParams)]
pub struct ListTasksQuery {
    pub status: Option<String>,
    pub r#type: Option<String>,
}

// ─── Handlers ────────────────────────────────────────────────────────────────

#[utoipa::path(
    get,
    path = "/tasks",
    tag = "Tasks",
    summary = "List tasks",
    description = "List tasks with optional status and type filters.",
    security(("Bearer" = [])),
    params(ListTasksQuery),
    responses(
        (status = 200, description = "Task list"),
        (status = 403, description = "Forbidden"),
    )
)]
pub async fn list_tasks(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    Extension(subscriber_counts): Extension<SubscriberCounts>,
    Query(query): Query<ListTasksQuery>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::EventSubscribe,
        None,
    ) {
        return Err(AppError::Forbidden);
    }

    let mut filter = TaskFilter::default();

    if let Some(ref status_str) = query.status {
        let statuses: Vec<TaskStatus> = status_str
            .split(',')
            .filter(|s| !s.is_empty())
            .filter_map(|s| serde_json::from_value(serde_json::Value::String(s.to_string())).ok())
            .collect();
        if !statuses.is_empty() {
            filter.status = Some(statuses);
        }
    }

    if let Some(ref type_str) = query.r#type {
        filter.types = Some(vec![type_str.clone()]);
    }

    let tasks = engine.list_tasks(filter).await?;
    let mut enriched = Vec::with_capacity(tasks.len());
    for task in &tasks {
        let subscriber_count = get_subscriber_count(&subscriber_counts, &task.id).await;
        let mut task_json = serde_json::to_value(task).unwrap();
        if let Some(obj) = task_json.as_object_mut() {
            obj.insert("hot".to_string(), json!(subscriber_count > 0));
            obj.insert("subscriberCount".to_string(), json!(subscriber_count));
        }
        enriched.push(task_json);
    }

    Ok(axum::Json(json!({ "tasks": enriched })))
}

#[utoipa::path(
    post,
    path = "/tasks",
    tag = "Tasks",
    summary = "Create a new task",
    security(("Bearer" = [])),
    request_body = CreateTaskBody,
    responses(
        (status = 201, description = "Task created", body = taskcast_core::Task),
        (status = 400, description = "Validation error"),
        (status = 403, description = "Forbidden"),
    )
)]
pub async fn create_task(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    axum::Json(body): axum::Json<CreateTaskBody>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::TaskCreate,
        None,
    ) {
        return Err(AppError::Forbidden);
    }

    let input = CreateTaskInput {
        id: body.id,
        r#type: body.r#type,
        params: body.params,
        metadata: body.metadata,
        ttl: body.ttl,
        webhooks: body.webhooks,
        cleanup: body.cleanup,
        auth_config: body.auth_config,
        tags: body.tags,
        assign_mode: body.assign_mode,
        cost: body.cost,
        disconnect_policy: body.disconnect_policy,
    };

    let task = engine.create_task(input).await?;
    Ok((StatusCode::CREATED, axum::Json(task)))
}

#[utoipa::path(
    get,
    path = "/tasks/{task_id}",
    tag = "Tasks",
    summary = "Get task by ID",
    security(("Bearer" = [])),
    params(("task_id" = String, Path, description = "Task ID")),
    responses(
        (status = 200, description = "Task details", body = taskcast_core::Task),
        (status = 404, description = "Task not found"),
        (status = 403, description = "Forbidden"),
    )
)]
pub async fn get_task(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    Extension(subscriber_counts): Extension<SubscriberCounts>,
    Path(task_id): Path<String>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::EventSubscribe,
        Some(&task_id),
    ) {
        return Err(AppError::Forbidden);
    }

    let task = engine
        .get_task(&task_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Task not found".to_string()))?;

    let subscriber_count = get_subscriber_count(&subscriber_counts, &task_id).await;
    let mut task_json = serde_json::to_value(&task).unwrap();
    if let Some(obj) = task_json.as_object_mut() {
        obj.insert("hot".to_string(), json!(subscriber_count > 0));
        obj.insert("subscriberCount".to_string(), json!(subscriber_count));
    }

    Ok(axum::Json(task_json))
}

#[utoipa::path(
    patch,
    path = "/tasks/{task_id}/status",
    tag = "Tasks",
    summary = "Transition task status",
    security(("Bearer" = [])),
    params(("task_id" = String, Path, description = "Task ID")),
    request_body = TransitionBody,
    responses(
        (status = 200, description = "Updated task", body = taskcast_core::Task),
        (status = 400, description = "Invalid transition"),
        (status = 404, description = "Task not found"),
        (status = 403, description = "Forbidden"),
    )
)]
pub async fn transition_task(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    Path(task_id): Path<String>,
    axum::Json(body): axum::Json<TransitionBody>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::TaskManage,
        Some(&task_id),
    ) {
        return Err(AppError::Forbidden);
    }

    let payload = if body.result.is_some()
        || body.error.is_some()
        || body.reason.is_some()
        || body.ttl.is_some()
        || body.resume_after_ms.is_some()
        || body.blocked_request.is_some()
    {
        let error = body.error.map(|e| TaskError {
            code: e.code,
            message: e.message,
            details: e.details,
        });
        Some(TransitionPayload {
            result: body.result,
            error,
            reason: body.reason,
            ttl: body.ttl,
            resume_after_ms: body.resume_after_ms,
            blocked_request: body.blocked_request,
        })
    } else {
        None
    };

    let task = engine
        .transition_task(&task_id, body.status, payload)
        .await
        .map_err(|e| match &e {
            EngineError::TaskNotFound(_) => AppError::NotFound(e.to_string()),
            EngineError::InvalidTransition { .. } => AppError::Engine(e),
            EngineError::TaskTerminal(_) => AppError::BadRequest(e.to_string()),
            _ => AppError::Engine(e),
        })?;

    Ok(axum::Json(task))
}

#[utoipa::path(
    post,
    path = "/tasks/{task_id}/events",
    tag = "Events",
    summary = "Publish events to a task",
    description = "Supports single event or batch (array) publishing.",
    security(("Bearer" = [])),
    params(("task_id" = String, Path, description = "Task ID")),
    responses(
        (status = 201, description = "Events published"),
        (status = 400, description = "Validation error"),
        (status = 404, description = "Task not found"),
        (status = 403, description = "Forbidden"),
    )
)]
pub async fn publish_events(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    Path(task_id): Path<String>,
    axum::Json(body): axum::Json<serde_json::Value>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::EventPublish,
        Some(&task_id),
    ) {
        return Err(AppError::Forbidden);
    }

    let is_batch = body.is_array();

    let inputs: Vec<PublishEventBody> = if is_batch {
        serde_json::from_value(body)
            .map_err(|e| AppError::BadRequest(e.to_string()))?
    } else {
        let single: PublishEventBody = serde_json::from_value(body)
            .map_err(|e| AppError::BadRequest(e.to_string()))?;
        vec![single]
    };

    let mut events = Vec::new();
    for input in inputs {
        let event_input = PublishEventInput {
            r#type: input.r#type,
            level: input.level,
            data: input.data,
            series_id: input.series_id,
            series_mode: input.series_mode,
            series_acc_field: input.series_acc_field,
        };
        let event = engine
            .publish_event(&task_id, event_input)
            .await
            .map_err(|e| match &e {
                EngineError::TaskNotFound(_) => AppError::NotFound(e.to_string()),
                EngineError::TaskTerminal(_) => AppError::BadRequest(e.to_string()),
                _ => AppError::Engine(e),
            })?;
        events.push(serde_json::to_value(&event).unwrap());
    }

    let body = if is_batch {
        json!(events)
    } else {
        events.into_iter().next().unwrap()
    };

    Ok((StatusCode::CREATED, axum::Json(body)))
}

#[utoipa::path(
    get,
    path = "/tasks/{task_id}/events/history",
    tag = "Events",
    summary = "Query event history",
    security(("Bearer" = [])),
    params(("task_id" = String, Path, description = "Task ID"), HistoryQuery),
    responses(
        (status = 200, description = "Event list", body = Vec<taskcast_core::TaskEvent>),
        (status = 404, description = "Task not found"),
        (status = 403, description = "Forbidden"),
    )
)]
pub async fn get_event_history(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    Path(task_id): Path<String>,
    Query(query): Query<HistoryQuery>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::EventHistory,
        Some(&task_id),
    ) {
        return Err(AppError::Forbidden);
    }

    // Check task exists
    let _task = engine
        .get_task(&task_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Task not found".to_string()))?;

    let since = if query.since_id.is_some()
        || query.since_index.is_some()
        || query.since_timestamp.is_some()
    {
        Some(SinceCursor {
            id: query.since_id,
            index: query.since_index,
            timestamp: query.since_timestamp,
        })
    } else {
        None
    };

    let opts = if since.is_some() || query.limit.is_some() {
        Some(EventQueryOptions {
            since,
            limit: query.limit,
        })
    } else {
        None
    };

    let mut events = engine.get_events(&task_id, opts).await?;

    let series_format = query.series_format.as_deref().unwrap_or("delta");
    if series_format == "accumulated" {
        let engine_ref = Arc::clone(&engine);
        events = taskcast_core::series::collapse_accumulate_series(
            &events,
            |tid: &str, sid: &str| {
                let eng = Arc::clone(&engine_ref);
                let tid = tid.to_string();
                let sid = sid.to_string();
                async move {
                    eng.get_series_latest(&tid, &sid)
                        .await
                        .map_err(|e| Box::new(e) as Box<dyn std::error::Error + Send + Sync>)
                }
            },
        )
        .await
        .unwrap_or(events);
    }

    Ok(axum::Json(events))
}

// ─── Resolve / Request Handlers ─────────────────────────────────────────────

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolveBody {
    pub data: serde_json::Value,
}

pub async fn resolve_task(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    Path(task_id): Path<String>,
    axum::Json(body): axum::Json<ResolveBody>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::TaskResolve,
        Some(&task_id),
    ) {
        return Err(AppError::Forbidden);
    }

    let task = engine
        .get_task(&task_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Task not found".to_string()))?;

    if task.status != TaskStatus::Blocked {
        return Err(AppError::BadRequest("Task is not blocked".to_string()));
    }

    let result = if body.data.is_object() {
        let map: HashMap<String, serde_json::Value> =
            serde_json::from_value(body.data).unwrap_or_default();
        Some(map)
    } else {
        let mut map = HashMap::new();
        map.insert("resolution".to_string(), body.data);
        Some(map)
    };

    let updated = engine
        .transition_task(
            &task_id,
            TaskStatus::Running,
            Some(TransitionPayload {
                result,
                ..Default::default()
            }),
        )
        .await
        .map_err(|e| AppError::BadRequest(e.to_string()))?;

    Ok(axum::Json(updated))
}

pub async fn get_blocked_request(
    State(engine): State<Arc<TaskEngine>>,
    Extension(auth): Extension<AuthContext>,
    Path(task_id): Path<String>,
) -> Result<impl IntoResponse, AppError> {
    if !check_scope(
        &auth,
        taskcast_core::PermissionScope::TaskResolve,
        Some(&task_id),
    ) {
        return Err(AppError::Forbidden);
    }

    let task = engine
        .get_task(&task_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Task not found".to_string()))?;

    if task.status != TaskStatus::Blocked {
        return Err(AppError::NotFound("No blocked request".to_string()));
    }

    match task.blocked_request {
        Some(request) => Ok(axum::Json(serde_json::to_value(request).unwrap())),
        None => Err(AppError::NotFound("No blocked request".to_string())),
    }
}
